### ISSUE TITLE
uncaught exceptions in the networking stack

### ISSUE THREAD

**Issue by rixed** [mirage/mirage#601]
I have a (XEN) program that dies (uncaught exception, exit 2) when the peer it
is connected to disappear.  It will first retransmit some TCP segments (`TCP
retransmission on timer seq ...`) and then crash:

```
Fatal error: exception Ipv4.Make(Ethif)(Arpv4).Routing.No_route_to_destination_address(_)
Mirage exiting with status 2
```

The exception is thrown from some asynchronous thread spawned by
mirageos itself. Indeed, it seems the networking stack rely a lot on the
Lwt.async call without always trying to catch the exceptions.
I understand that the TCP stack will encounter errors asynchronously, but I
believe the unix model of error reporting is more convenient: the error should be saved in the "socket" internal state and be reported (synchronously) next time the socket is used for
anything.

Currently to avoid the termination of the program one can set the Lwt `async_exception_hook` but it is passed only the faulty IP; the channels depending on this IP still have to be closed.

(Note: I would have written to mirageos-devel@lists.xenproject.org instead of opening a bug but the ML still drops my emails it seems)

**Comment by hannesm** [comment]
see also https://github.com/mirage/mirage-tcpip/issues/240 ... and https://github.com/mirage/mirage/issues/506 .. I don't think someone is actively solving this atm

**Comment by yomimono** [comment]
This is fixed by the just-merged #711 and companion commit in `tcpip` at https://github.com/mirage/mirage-tcpip/pull/269 .
